modules/autotagging-commons/common/src/main/java/org/opensearch/rule/SecurityAttribute.java

```diff
@@ -0,0 +1,137 @@
+     *     "role": ["all_access"]
+     * }
+     * will be parsed into a set with values "username|alice" and "role|all_access"
+     * @param parser the XContent parser
+     */
+    @Override
+    public Set<String> fromXContentParseAttributeValues(XContentParser parser) throws IOException {
+        Set<String> resultSet = new HashSet<>();
+
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
+            throw new XContentParseException(
+                parser.getTokenLocation(),
+                "Expected START_OBJECT token for " + getName() + " attribute but got " + parser.currentToken()
+            );
+        }
+        List<String> allowedSubfieldsName = PRIORITIZED_SUBFIELDS.values().stream().map(Attribute::getName).toList();
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            String subFieldName = parser.currentName();
+            parser.nextToken();
+            if (!allowedSubfieldsName.contains(subFieldName)) {
+                throw new XContentParseException(
+                    parser.getTokenLocation(),
+                    "Invalid field: " + subFieldName + ". Allowed fields are: " + String.join(", ", allowedSubfieldsName)
+                );
+            }
+            if (parser.currentToken() != XContentParser.Token.START_ARRAY) {
+                throw new XContentParseException(
+                    parser.getTokenLocation(),
+                    "Expected array for field: " + subFieldName + " but got " + parser.currentToken()
+                );
+            }
+            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                if (parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+                    // prefix each value with the subFieldName (e.g., "username|name1")
+                    resultSet.add(String.join("|", subFieldName, parser.text()));
+                } else {
+                    throw new XContentParseException(
+                        parser.getTokenLocation(),
+                        "Expected string value in array under '" + subFieldName + "', but got " + parser.currentToken()
+                    );
+                }
+            }
+        }
+
+        return resultSet;
+    }
+
+    @Override
+    public void toXContentWriteAttributeValues(XContentBuilder builder, Set<String> values) throws IOException {
+        builder.startObject(getName());
+        Map<String, Set<String>> grouped = new HashMap<>();
+
+        // For each string in the values set, split it into two parts using the first underscore as delimiter:
+        // parts[0] is the prefix (e.g., "username" or "role")
+        // parts[1] is the actual value (e.g., "name1", "role1")
+        for (String value : values) {
+            String[] parts = value.split("\\|", 2);
+            if (parts.length == 2) {
+                grouped.computeIfAbsent(parts[0], k -> new HashSet<>()).add(parts[1]);
+            }
+        }
```

Comment:
I think encapsulating the encoding and decoding logic into a separate construct will make more sense since that logic will not be just reusable but separate out the testing as well